File: WinterQuarterCS60/HomeWork1/wordle.cpp
// TODO: When finished, write a file comment that describes the program.
// Write your name and date, and cite anyone who helped you.
/*
Sami Shibli January 15 2023
This program stimulates a simple version
of the webgame wordle
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <ctime>
#include <cstring>

using namespace std;

// program constants
const string FILENAME = "wordbank.txt";
const int NUM_LETTERS = 5;
const int NUM_GUESSES = 6;

// helper function declarations
string getWord();
string getGuess();
void processGuess(string guess, string secret);

// Main plays the game and report the outcome
int main()
{

	string secret = getWord();
	cout << "The secret word has been selected. Start guessing!" << endl;
	string guess;

	for (int i = 0; i < NUM_GUESSES; i++)
	{
		cout << "Guess # " << (i + 1) << ": " << endl;
		guess = getGuess();

		for (int j = 0; j < guess.length(); j++)
		{
			guess[j] = toupper(guess[j]);
		}

		if (guess == secret)
		{
			break;
		}
		processGuess(guess, secret);
	}

	if (guess == secret)
	{
		cout << "You guessed the word!" << endl;
	}
	else
	{
		cout << "You ran out of guesses before guessing the word: " << secret << endl; // for testing; remove before submitting

		/* TODO: write code to play the game and report the outcome here */

		return 0;
	}
}

// picks a random word from the specified file
string getWord()
{
	// open the file
	ifstream f(FILENAME);
	// read words line-by-line into a vector (array-like data structure)
	vector<string> allWords;
	string word;
	while (getline(f, word))
	{
		allWords.push_back(word);
	}
	// close the file
	f.close();
	// pick a random word from the vector and capitalize its letters
	srand(time(nullptr));
	word = allWords[rand() % allWords.size()];
	for (int i = 0; i < word.length(); i++)
	{
		word[i] = toupper(word[i]);
	}
	return word;
}

// solicit a guess from the user. Requires
string getGuess()
{
	string guess;
	cout << "Enter in a guess!" << endl;
	cin >> guess;
	while (guess.size() != NUM_LETTERS)
	{
		cout << "your guess must be 5 letters" << endl;
		cout << "Enter in another guess!";
		cin >> guess;
	}
	return guess;
}
// check to see letter by letter if its in the correct spot.
void processGuess(string guess, string secret)
{
	for (int i = 0; i < NUM_LETTERS; i++)
	{
		if (secret[i] == guess[i])
		{
			cout << guess[i] << " is in position " << i + 1 << endl;
		}
		else
		{
			bool in_word = false;
			for (int j = 0; j < NUM_LETTERS; j++)
			{
				if (secret[j] == guess[i])
				{
					in_word = true;
					break;
				}
			}
			if (in_word)
			{
				cout << guess[i] << " is in the word but not in position " << i + 1 << endl;
			}
			else
			{
				cout << guess[i] << " is not in the word" << endl;
			}
		}
	}
}

File: WinterQuarterCS60/HomeWork1/wordle.py
# wordle.py is a program for a simplified version of the word game wordle.
# This is fully implemented in python, and you should use it as a reference
# for HW1, for which you should provide a new implementation in C++.
# Sara Krehbiel, 1/9/23

import random

# open word back file and pick and return a random word from it
def get_word():
	f = open(filename)
	all_words = f.readlines()
	f.close()
	word = all_words[random.randrange(len(all_words))].strip().upper()
	return word

# solicit a guess from a user; require 5 letters and capitalize automatically
def get_guess(prompt):
	guess = input(prompt).upper()
	while len(guess) != num_letters:
		print('Your guess must be',num_letters,'letters.')
		guess = input(prompt).upper()
	return guess

# report the letter-by-letter similarity of a guess to a secret word
def process_guess(guess, secret):
	for i in range(num_letters):
		letter = guess[i]
		# check whether the ith letter is in the correct spot
		if secret[i] == letter:
			print(letter,'is in position',i+1)
		else:
			# check whether the ith letter is (elsewhere) in the secret
			in_word = False
			for j in range(num_letters):
				if secret[j] == letter:
					in_word = True
					break
			if in_word:
				print(letter,'is in the word but not in position',i+1)
			else:
				print(letter,'is not in the word')

# specify word back filename, word length, and max number of guesses
filename = 'wordbank.txt'
num_letters = 5
num_guesses = 6

# pick secret word
secret = get_word()
#secret = 'PLANE' # feel free to hardcode your own words when testing
#print(secret) # you may also print out the word for testing
print('The secret word has been selected. Start guessing!')

# play game
for i in range(num_guesses):
	guess = get_guess('Guess #' + str(i+1) + ': ')
	if guess == secret: break
	process_guess(guess,secret)

# report outcome
if guess == secret:
	print('You guessed the word!')
else:
	print('You ran out of guesses before guessing the word:',secret)
